internal/k8s/controllers/config_controller_test.go

```diff
@@ -0,0 +1,279 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	v1beta1 "go.universe.tf/metallb/api/v1beta1"
+	v1beta2 "go.universe.tf/metallb/api/v1beta2"
+	"go.universe.tf/metallb/internal/config"
+	metallbcfg "go.universe.tf/metallb/internal/config"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func TestConfigController(t *testing.T) {
+	tests := []struct {
+		desc                    string
+		handlerRes              SyncState
+		validResources          bool
+		expectReconcileFails    bool
+		expectForceReloadCalled bool
+	}{
+		{
+			desc:                    "handler returns SyncStateSuccess, valid resources",
+			handlerRes:              SyncStateSuccess,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateError, valid resources",
+			handlerRes:              SyncStateError,
+			validResources:          true,
+			expectReconcileFails:    true,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateErrorNoRetry, valid resources",
+			handlerRes:              SyncStateErrorNoRetry,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateReprocessAll, valid resources",
+			handlerRes:              SyncStateReprocessAll,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: true,
+		},
+		{
+			desc:                    "handler returns SyncStateSuccess, invalid resources",
+			handlerRes:              SyncStateSuccess,
+			validResources:          false,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+	}
+	for _, test := range tests {
+		var resources metallbcfg.ClusterResources
+		if test.validResources {
+			resources = configControllerValidResources
+		} else {
+			resources = configControllerInvalidResources
+		}
+
+		fakeClient := newFakeClient(t, resources)
+
+		expectedCfg, err := config.For(resources, config.DontValidate)
+		if err != nil && test.validResources {
+			t.Fatalf("test %s failed, got unexpected error: %v", test.desc, err)
```

Comment:
same here, maybe saying it failed to create the config?

Reply:
done